chore(ci): optimize runner costs

webjunkie opened the pull request:
## Summary
Reduce Depot runner costs by right-sizing runners and moving non-Docker jobs to free GitHub runners.

## Changes
- **Python linting**: `depot-ubuntu-latest` → `ubuntu-latest` (ruff/mypy don't need Docker cache)
- **TypeScript checks**: `depot-ubuntu-24.04-8` → `depot-ubuntu-24.04-4` (tsc is single-threaded, wasting 4 cores)
- **E2E Playwright**: `depot-ubuntu-latest-8` → `depot-ubuntu-latest-4` (uses 3 workers per playwright.config.ts, doesn't need 8 cores)
- **Frontend path detection**: `depot-ubuntu-24.04-small` → `ubuntu-latest` (just path filtering)
- **Frontend telemetry**: `depot-ubuntu-24.04` → `ubuntu-latest` (just API calls)

## Performance impact
Measured across multiple branches before/after changes:

**TypeScript checks:** 3m50s avg (8-core) → 3m49s (4-core) - **no impact**
**Python linting:** 3m47s avg (Depot) → 5m25s (free) - **+1m38s** (non-blocking, runs in parallel)
**E2E Playwright:** ~11m (4-core) - **similar to typical full runs**

## Expected savings
- TypeScript + E2E: ~200k billable minutes/month (4x → 2x multiplier = 50% cost cut)
- Python linting: 100% savings (moved to free runners)

## Test plan
- [x] CI passes on this PR
- [ ] Monitor runner performance over next few days
- [ ] Check actual billing impact after a week

webjunkie committed:
chore(ci): optimize runner costs

Reduce Depot runner costs by right-sizing runners and moving non-Docker jobs to free GitHub runners.

Changes:
- Python linting: depot-ubuntu-latest → ubuntu-latest (no Docker needed)
- TypeScript checks: depot-ubuntu-24.04-8 → depot-ubuntu-24.04-4 (tsc is single-threaded, wasting 4 cores)
- E2E Playwright: depot-ubuntu-latest-8 → depot-ubuntu-latest-4 (uses 3 workers, doesn't need 8 cores)
- Frontend path detection: depot-ubuntu-24.04-small → ubuntu-latest (just path filtering)
- Frontend telemetry: depot-ubuntu-24.04 → ubuntu-latest (just API calls)

Expected savings: ~200k billable minutes/month from TypeScript + E2E downgrades alone (4x → 2x multiplier).

joshsny reviewed (approved):
assuming we've checked the impact of these on the jobs individually, looks good - left some comments from when I last tested things

webjunkie committed:
chore(ci): use depot runner for Python type checks

Reverts mypy check to depot runner since mypy is single-threaded and benefits from faster CPUs. Free runners are too slow for type checking.

webjunkie commented on .github/workflows/ci-e2e-playwright.yml:
```diff
@@ -259,7 +259,7 @@ jobs:
         name: Playwright E2E tests
         needs: [changes]
-        runs-on: depot-ubuntu-latest-8
+        runs-on: depot-ubuntu-latest-4
```
Good point, this workflow didn't even run here due to skip conditions 😓

webjunkie committed:
test: trigger playwright run to compare 4-core vs 8-core performance

webjunkie commented on .github/workflows/ci-python.yml:
```diff
@@ -40,7 +40,7 @@ jobs:
 
         name: Python code quality checks
-        runs-on: depot-ubuntu-latest
+        runs-on: ubuntu-latest
```
Agreed. Reverted this one.

webjunkie committed:
Merge branch 'master' into optimize-ci-runner-costs

# Conflicts:
#	.github/workflows/ci-e2e-playwright.yml
#	.github/workflows/ci-frontend.yml

webjunkie committed:
revert(ci): restore 8 CPUs for Playwright E2E tests

Statistical analysis showed:
- 4 CPUs: 697.5s average (44% cost savings)
- 8 CPUs: 620.5s baseline
- Impact: +77s per run (12.4% slower)

Keep TypeScript checks at 4 CPUs (46% savings, no perf impact).
Revert Playwright to 8 CPUs to maintain fast feedback loops.

webjunkie committed:
perf(ci): use ubuntu-latest for Node.js changes detection

Similar to frontend CI, the changes detection job is git-heavy and benefits
from GitHub's infrastructure proximity. Saves Depot runner costs with better
performance for lightweight git operations.
